cohttp: revert change to Header.replace, add Header.update

mseri opened the pull request:
A change in #694 modified the semantics of Header.replace.
The semantics change is reverted, and a new Header.update function is introduced which changes a header value only when present.

Signed-off-by: Marcello Seri <marcello.seri@gmail.com>

dinosaure commented:
May be we should add some unit tests to avoid such breakage.

mseri commented:
Thanks @dinosaure. Before merging, I wanted to ask if it is better to have `update` do nothing when the key is absent (like here) or fail like suggested at the issue page: https://github.com/mirage/ocaml-cohttp/issues/701

mseri commented:
> May be we should add some unit tests to avoid such breakage.

I agree. Once we settle on a semantic for `update`, I'll add all the tests

dinosaure commented:
According #694:
> The replace function was not working properly (it should do nothing if the key is not already in the headers). I have also corrected some minor typos in documentation.

I think a case such as a use of `Header.replace` with a `fold` (eg. `List.fold_left`) should not fail and it should ensure that we inserted all required fields (even if they already exist at the beginning).

PS: more deeply, I think the use of a `StringMap` is not so good and a simple associative list should be better - but it's my opinion.

mseri commented:
I don't want to change the underlying representation at this stage, but I would not mind an issue to discuss it for cohttp 3.0.

Before discussing a new API or a new internal description, I would like to restore the semantics and add an `update` option. I think the suggestion by @anil on the issue page of returning a `'header option` instead of silently doing nothing is better than what I do in the PR. I would actually avoid producing the `_exn` version, interested users can easily add their exception raising on None.

mseri commented:
I have changed the function name to `update_existing`, added (I hope clear) descriptions and added unit tests for the expected behaviours. The CI failures seem due to the usual extra pedantic check of the logs if I am not mistaken.

dinosaure commented:
The error is about `async` and it is unrelated to `cohttp`:
```
#18 9.217 (cd _build/default/cohttp-async/test && ./test_async_integration.exe)
#18 9.217 Uncaught exception:
#18 9.217   
#18 9.217   (Sys_error "/etc/localtime: No such file or directory")
#18 9.217 
#18 9.217 Raised by primitive operation at file "stdlib.ml", line 398, characters 28-54
#18 9.217 Called from file "src/zone.ml", line 348, characters 21-49
#18 9.217 Called from file "src/timezone.ml", line 142, characters 8-75
#18 9.217 Called from file "camlinternalLazy.ml", line 31, characters 17-27
#18 9.217 Re-raised at file "camlinternalLazy.ml", line 36, characters 4-11
#18 9.217 Called from file "src/core_time_ns.ml", line 462, characters 38-56
#18 9.217 Called from file "src/raw_scheduler.ml", line 895, characters 41-69
#18 9.217 Called from file "cohttp-async/test/test_async_integration.ml", line 122, characters 21-49
```

So, CI should be happy :+1:.

mseri commented:
I modified `Header.update` to follow more closely the semantics of `Map.update` as suggested by @pm-mck. Instead of using the more general `f: string list option -> string list option`, I followed the `get` function and limited the type to `f: string option -> string option`, with some documented concatenation and splitting under the table.

If you feel that using a more general type, say `f: string list option -> string list option` or `f: string list -> string list` and reduce the magic in the function, I'm happy to modify it, otherwise I think this is ready to merge.

The version I wrote only accepting a `string list` was messier in my opinion (I was associating `[]` and `None` in that case), the one accepting `string list option` allows users to associate list of values to headers that should not accept them (note that you can still inject lists of values using `map`).

mseri commented:
@dinosaure can you have a look at the latest changes? If you are happy with it, I'd merge and draft a new minor release

mseri commented:
Thanks!
